Lib/test/test_locale_coercion.py

```diff
@@ -0,0 +1,199 @@
+        result, py_cmd = run_python_until_end(
+            "-c", cmd,
+            __isolated=True,
+            **env_vars
+        )
+        if not result.rc == 0:
+            result.fail(py_cmd)
+        # All subprocess outputs in this test case should be pure ASCII
+        child_fsencoding = result.out.decode("ascii").rstrip()
+        child_stderr_lines = result.err.decode("ascii").rstrip().splitlines()
+        return child_fsencoding, child_stderr_lines
+
+
+    def test_C_utf8_locale(self):
+        # Ensure the C.UTF-8 locale is accepted entirely without complaint
+        base_var_dict = {
+            "LANG": "",
+            "LC_CTYPE": "",
+            "LC_ALL": "",
+        }
+        for env_var in ("LC_ALL", "LC_CTYPE", "LANG"):
+            with self.subTest(env_var=env_var):
+                var_dict = base_var_dict.copy()
+                var_dict[env_var] = "C.UTF-8"
+                fsencoding, stderr_lines = self._get_child_fsencoding(var_dict)
+                self.assertEqual(fsencoding, "utf-8")
+                self.assertFalse(stderr_lines)
+
+
+    def _check_c_locale_coercion(self, expected_fsencoding, coerce_c_locale):
+        """Check the handling of the C locale for various configurations
+
+        Parameters:
+            expected_fsencoding: the encoding the child is expected to report
+            allow_c_locale: setting to use for PYTHONALLOWCLOCALE
+              None: don't set the variable at all
+              str: the value set in the child's environment
+        """
+        expected_warning = []
+        if coerce_c_locale != "0":
+            # Check C locale is coerced with a warning on stderr
+            expected_warning.append(self.EXPECTED_COERCION_WARNING)
+        base_var_dict = {
+            "LANG": "",
+            "LC_CTYPE": "",
+            "LC_ALL": "",
+        }
+        for env_var in ("LC_ALL", "LC_CTYPE", "LANG"):
+            for locale_to_set in ("", "C", "POSIX", "invalid.ascii"):
+                with self.subTest(env_var=env_var,
+                                  nominal_locale=locale_to_set,
+                                  PYTHONCOERCECLOCALE=coerce_c_locale):
+                    var_dict = base_var_dict.copy()
+                    var_dict[env_var] = locale_to_set
+                    if coerce_c_locale is not None:
+                        var_dict["PYTHONCOERCECLOCALE"] = coerce_c_locale
+                    fsencoding, stderr_lines = self._get_child_fsencoding(var_dict)
+                    self.assertEqual(fsencoding, expected_fsencoding)
+                    self.assertEqual(stderr_lines, expected_warning)
+
+
+    def test_test_PYTHONCOERCECLOCALE_not_set(self):
+        # This should coerce to the C.UTF-8 locale by default
+        self._check_c_locale_coercion("utf-8", coerce_c_locale=None)
+
+    def test_PYTHONCOERCECLOCALE_not_zero(self):
+        # *Any* string other that "0" is considered "set" for our purposes
```

Comment:
Docs fixed in https://github.com/python/cpython/pull/659/commits/1c3a2706ef276ca66f676218395b9d792ef3ca0e